File: packages/rust/tests/tx.rs
//! Native + hybrid (classical + post-quantum) transaction tests.
//!
//! `bank_send` and `build_hybrid_tx` are pure (no network); `broadcast` is tested
//! against a tiny local mock HTTP server that records the request, mirroring the
//! query tests. No real node or network access is used.

use std::convert::Infallible;
use std::sync::{Arc, Mutex};

use http_body_util::{BodyExt, Full};
use hyper::body::Bytes;
use hyper::service::service_fn;
use hyper::{Request, Response};
use hyper_util::rt::TokioIo;
use tokio::net::TcpListener;

use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine;
use cosmrs::proto::cosmos::bank::v1beta1::MsgSend;
use cosmrs::proto::cosmos::tx::v1beta1::{SignDoc, TxBody, TxRaw};
use cosmrs::proto::traits::Message;

use qorechain::accounts::derive_native_account;
use qorechain::pqc::{
    generate_pqc_keypair, pqc_verify, HYBRID_SIG_TYPE_URL, MLDSA87_SIGNATURE_LEN,
};
use qorechain::tx::{
    bank_send, broadcast, build_hybrid_tx, fee_from_estimate, BankSendParams, BroadcastMode,
    BuildHybridTxParams, Coin, Fee, Message as TxMessage,
};

/// Public test mnemonic only (BIP-39 test vector); never a real secret.
const TEST_MNEMONIC: &str =
    "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about";
const TEST_CHAIN_ID: &str = "qorechain-diana";
const MSG_SEND_TYPE_URL: &str = "/cosmos.bank.v1beta1.MsgSend";

fn sample_fee() -> Fee {
    Fee {
        amount: vec![Coin {
            denom: "uqor".into(),
            amount: "5000".into(),
        }],
        gas: "200000".into(),
        granter: String::new(),
        payer: String::new(),
    }
}

/// A big-endian 4-byte length prefix, matching the chain contract framing.
fn be32(n: u32) -> [u8; 4] {
    n.to_be_bytes()
}

/// Re-create the PQC framing `BE32(len(b0)) || b0 || BE32(len(a)) || a`.
fn frame(b0: &[u8], a: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(8 + b0.len() + a.len());
    out.extend_from_slice(&be32(b0.len() as u32));
    out.extend_from_slice(b0);
    out.extend_from_slice(&be32(a.len() as u32));
    out.extend_from_slice(a);
    out
}

// --- mock HTTP server (mirrors tests/query.rs) ---

#[derive(Debug, Clone)]
struct Recorded {
    method: String,
    path: String,
    body: String,
}

struct MockServer {
    base_url: String,
    recorded: Arc<Mutex<Vec<Recorded>>>,
    _shutdown: tokio::sync::oneshot::Sender<()>,
}

impl MockServer {
    async fn start(response_body: &'static str) -> Self {
        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();
        let base_url = format!("http://{addr}");
        let recorded: Arc<Mutex<Vec<Recorded>>> = Arc::new(Mutex::new(Vec::new()));
        let (tx, mut rx) = tokio::sync::oneshot::channel::<()>();

        let rec = recorded.clone();
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    _ = &mut rx => break,
                    accepted = listener.accept() => {
                        let (stream, _) = match accepted {
                            Ok(v) => v,
                            Err(_) => continue,
                        };
                        let io = TokioIo::new(stream);
                        let rec = rec.clone();
                        tokio::spawn(async move {
                            let svc = service_fn(move |req: Request<hyper::body::Incoming>| {
                                let rec = rec.clone();
                                async move {
                                    let method = req.method().to_string();
                                    let path = req.uri().path().to_string();
                                    let bytes =
                                        req.into_body().collect().await.unwrap().to_bytes();
                                    let body = String::from_utf8_lossy(&bytes).to_string();
                                    rec.lock().unwrap().push(Recorded { method, path, body });
                                    Ok::<_, Infallible>(Response::new(Full::new(Bytes::from(
                                        response_body,
                                    ))))
                                }
                            });
                            let _ = hyper::server::conn::http1::Builder::new()
                                .serve_connection(io, svc)
                                .await;
                        });
                    }
                }
            }
        });

        MockServer {
            base_url,
            recorded,
            _shutdown: tx,
        }
    }

    fn last(&self) -> Recorded {
        self.recorded.lock().unwrap().last().cloned().unwrap()
    }
}

// --- fee helper ---

#[test]
fn fee_from_estimate_uses_suggested_fee() {
    let fee = fee_from_estimate(
        &serde_json::json!({"suggested_fee_uqor":"1234","estimated_blocks":2}),
        "200000",
    )
    .unwrap();
    assert_eq!(fee.gas, "200000");
    assert_eq!(fee.amount.len(), 1);
    assert_eq!(fee.amount[0].denom, "uqor");
    assert_eq!(fee.amount[0].amount, "1234");
}

#[test]
fn fee_from_estimate_accepts_json_number() {
    let fee = fee_from_estimate(&serde_json::json!({"suggested_fee_uqor":4200}), "100000").unwrap();
    assert_eq!(fee.amount[0].amount, "4200");
}

#[test]
fn fee_from_estimate_rejects_empty() {
    assert!(fee_from_estimate(&serde_json::json!({}), "200000").is_err());
    assert!(fee_from_estimate(&serde_json::json!({"suggested_fee_uqor":"0"}), "200000").is_err());
}

// --- bank_send ---

#[test]
fn bank_send_builds_signed_tx_raw() {
    let acc = derive_native_account(TEST_MNEMONIC, 0).unwrap();
    let to = "qor1recipient00000000000000000000000000000";

    let built = bank_send(BankSendParams {
        private_key: acc.private_key.clone(),
        public_key: acc.public_key.clone(),
        from_address: acc.address.clone(),
        to_address: to.into(),
        amount: vec![Coin {
            denom: "uqor".into(),
            amount: "1000".into(),
        }],
        chain_id: TEST_CHAIN_ID.into(),
        account_number: 7,
        sequence: 3,
        fee: sample_fee(),
        memo: "hello".into(),
        timeout_height: 0,
    })
    .unwrap();

    assert!(!built.tx_raw_bytes.is_empty());

    let tx_raw = TxRaw::decode(built.tx_raw_bytes.as_slice()).unwrap();
    assert_eq!(tx_raw.signatures.len(), 1);
    assert_eq!(tx_raw.signatures[0].len(), 64);

    let body = TxBody::decode(tx_raw.body_bytes.as_slice()).unwrap();
    assert_eq!(body.memo, "hello");
    assert_eq!(body.messages.len(), 1);
    assert_eq!(body.messages[0].type_url, MSG_SEND_TYPE_URL);

    let msg = MsgSend::decode(body.messages[0].value.as_slice()).unwrap();
    assert_eq!(msg.from_address, acc.address);
    assert_eq!(msg.to_address, to);
    assert_eq!(msg.amount.len(), 1);
    assert_eq!(msg.amount[0].denom, "uqor");
    assert_eq!(msg.amount[0].amount, "1000");

    // Classical signature verifies over SignDoc(body, authInfo, chain, accNum).
    let sign_doc = SignDoc {
        body_bytes: tx_raw.body_bytes.clone(),
        auth_info_bytes: tx_raw.auth_info_bytes.clone(),
        chain_id: TEST_CHAIN_ID.into(),
        account_number: 7,
    };
    let sig = k256::ecdsa::Signature::from_slice(&tx_raw.signatures[0]).unwrap();
    let vk = k256::ecdsa::VerifyingKey::from_sec1_bytes(&acc.public_key).unwrap();
    use k256::ecdsa::signature::Verifier;
    vk.verify(&sign_doc.encode_to_vec(), &sig)
        .expect("classical signature must verify");
}

// --- broadcast ---

#[tokio::test]
async fn broadcast_posts_to_rest_endpoint() {
    for (mode, want_mode) in [
        (BroadcastMode::Sync, "BROADCAST_MODE_SYNC"),
        (BroadcastMode::Async, "BROADCAST_MODE_ASYNC"),
        (BroadcastMode::Block, "BROADCAST_MODE_BLOCK"),
    ] {
        let server = MockServer::start(r#"{"tx_response":{"txhash":"ABC123","code":0}}"#).await;
        let tx_bytes = vec![0x01u8, 0x02, 0x03];
        let resp = broadcast(&server.base_url, &tx_bytes, mode).await.unwrap();
        let last = server.last();
        assert_eq!(last.method, "POST");
        assert_eq!(last.path, "/cosmos/tx/v1beta1/txs");
        let payload: serde_json::Value = serde_json::from_str(&last.body).unwrap();
        assert_eq!(payload["mode"], want_mode);
        assert_eq!(payload["tx_bytes"], BASE64.encode(&tx_bytes));
        assert_eq!(resp["tx_response"]["txhash"], "ABC123");
    }
}

#[tokio::test]
async fn broadcast_http_error_is_surfaced() {
    // The mock always answers 200; for an error path we point at a closed port.
    let err = broadcast("http://127.0.0.1:1", &[0x01], BroadcastMode::Sync).await;
    assert!(err.is_err());
}

// --- hybrid PQC tx ---

#[test]
fn build_hybrid_tx_contract() {
    let acc = derive_native_account(TEST_MNEMONIC, 0).unwrap();
    let kp = generate_pqc_keypair().unwrap();

    let msg = MsgSend {
        from_address: acc.address.clone(),
        to_address: "qor1recipient00000000000000000000000000000".into(),
        amount: vec![cosmrs::proto::cosmos::base::v1beta1::Coin {
            denom: "uqor".into(),
            amount: "2500".into(),
        }],
    };

    let built = build_hybrid_tx(BuildHybridTxParams {
        private_key: acc.private_key.clone(),
        public_key: acc.public_key.clone(),
        pqc_secret_key: kp.secret_key.clone(),
        pqc_public_key: kp.public_key.clone(),
        messages: vec![TxMessage {
            type_url: MSG_SEND_TYPE_URL.into(),
            value: msg.encode_to_vec(),
        }],
        fee: sample_fee(),
        chain_id: TEST_CHAIN_ID.into(),
        account_number: 11,
        sequence: 5,
        memo: "pqc".into(),
        timeout_height: 0,
        include_pqc_public_key: false,
    })
    .unwrap();

    // ML-DSA-87 signature must be 4627 bytes and verify over the framed message.
    assert_eq!(built.pqc_signature.len(), MLDSA87_SIGNATURE_LEN);
    assert_eq!(MLDSA87_SIGNATURE_LEN, 4627);
    assert!(pqc_verify(
        &kp.public_key,
        &built.pqc_signed_message,
        &built.pqc_signature
    ));

    // Decode the final TxRaw + body.
    let tx_raw = TxRaw::decode(built.tx_raw_bytes.as_slice()).unwrap();
    assert_eq!(tx_raw.signatures.len(), 1);
    assert_eq!(tx_raw.signatures[0].len(), 64);

    let final_body = TxBody::decode(tx_raw.body_bytes.as_slice()).unwrap();

    // The PQC extension must be in extension_options (the CRITICAL slot).
    assert_eq!(final_body.extension_options.len(), 1);
    assert!(final_body.non_critical_extension_options.is_empty());
    let ext = &final_body.extension_options[0];
    assert_eq!(ext.type_url, HYBRID_SIG_TYPE_URL);

    // The Any.value must be the JSON shape.
    let ext_json: serde_json::Value = serde_json::from_slice(&ext.value).unwrap();
    assert_eq!(ext_json["algorithm_id"], 1);
    let sig_b64 = ext_json["pqc_signature"].as_str().unwrap();
    let sig_bytes = BASE64.decode(sig_b64).unwrap();
    assert_eq!(sig_bytes.len(), MLDSA87_SIGNATURE_LEN);
    // pqc_public_key omitted when not requested.
    assert!(ext_json.get("pqc_public_key").is_none());

    // KEY PROPERTY: strip the PQC ext from the final body, re-encode (B0'),
    // re-frame, and assert it equals the signed message.
    let mut stripped = final_body.clone();
    stripped.extension_options.clear();
    let b0_prime = stripped.encode_to_vec();
    let reframed = frame(&b0_prime, &tx_raw.auth_info_bytes);
    assert_eq!(reframed, built.pqc_signed_message);

    // A framing over the WITH-ext body must differ.
    let with_ext = frame(&tx_raw.body_bytes, &tx_raw.auth_info_bytes);
    assert_ne!(with_ext, built.pqc_signed_message);

    // The BE32 length prefix matches len(B0').
    let prefix = u32::from_be_bytes(built.pqc_signed_message[..4].try_into().unwrap());
    assert_eq!(prefix as usize, b0_prime.len());

    // Classical signature verifies over SignDoc(finalBody, A, chain, accNum).
    let sign_doc = SignDoc {
        body_bytes: tx_raw.body_bytes.clone(),
        auth_info_bytes: tx_raw.auth_info_bytes.clone(),
        chain_id: TEST_CHAIN_ID.into(),
        account_number: 11,
    };
    let vk = k256::ecdsa::VerifyingKey::from_sec1_bytes(&acc.public_key).unwrap();
    let sig = k256::ecdsa::Signature::from_slice(&tx_raw.signatures[0]).unwrap();
    use k256::ecdsa::signature::Verifier;
    vk.verify(&sign_doc.encode_to_vec(), &sig)
        .expect("classical signature must verify over final SignDoc");
}

#[test]
fn build_hybrid_tx_includes_public_key_when_requested() {
    let acc = derive_native_account(TEST_MNEMONIC, 0).unwrap();
    let kp = generate_pqc_keypair().unwrap();
    let msg = MsgSend {
        from_address: acc.address.clone(),
        to_address: "qor1recipient00000000000000000000000000000".into(),
        amount: vec![cosmrs::proto::cosmos::base::v1beta1::Coin {
            denom: "uqor".into(),
            amount: "1".into(),
        }],
    };
    let built = build_hybrid_tx(BuildHybridTxParams {
        private_key: acc.private_key.clone(),
        public_key: acc.public_key.clone(),
        pqc_secret_key: kp.secret_key.clone(),
        pqc_public_key: kp.public_key.clone(),
        messages: vec![TxMessage {
            type_url: MSG_SEND_TYPE_URL.into(),
            value: msg.encode_to_vec(),
        }],
        fee: sample_fee(),
        chain_id: TEST_CHAIN_ID.into(),
        account_number: 0,
        sequence: 0,
        memo: String::new(),
        timeout_height: 0,
        include_pqc_public_key: true,
    })
    .unwrap();

    let tx_raw = TxRaw::decode(built.tx_raw_bytes.as_slice()).unwrap();
    let final_body = TxBody::decode(tx_raw.body_bytes.as_slice()).unwrap();
    let ext_json: serde_json::Value =
        serde_json::from_slice(&final_body.extension_options[0].value).unwrap();
    let pk_b64 = ext_json["pqc_public_key"].as_str().unwrap();
    assert_eq!(BASE64.decode(pk_b64).unwrap(), kp.public_key);
}

File: packages/py/src/qorsdk/tx.py
"""Native transaction building, broadcast, and hybrid (classical + post-quantum)
signing for QoreChain.

This module mirrors the QoreChain TypeScript SDK's tx surface using ``cosmpy``'s
Cosmos protobuf types (``TxBody`` / ``AuthInfo`` / ``SignDoc`` / ``TxRaw`` /
``MsgSend``) and its secp256k1 keypair for SIGN_MODE_DIRECT signing. It provides:

- :func:`bank_send` — build and sign a ``MsgSend`` into a broadcast-ready
  ``TxRaw`` (classical secp256k1 only).
- :func:`broadcast` — POST signed tx bytes to the REST ``/cosmos/tx/v1beta1/txs``
  endpoint (modes ``sync`` / ``async`` / ``block``). Broadcasting needs a live
  node, so unit tests mock this HTTP POST.
- :func:`build_hybrid_tx` — assemble a hybrid transaction carrying a classical
  secp256k1 signature in ``TxRaw.signatures`` PLUS an ML-DSA-87 (Dilithium-5)
  signature attached to the ``TxBody`` as a ``PQCHybridSignature`` extension.

──────────────────────────────────────────────────────────────────────────────
 The wallet ↔ chain hybrid contract (enforced by the chain; matches the TS SDK)
──────────────────────────────────────────────────────────────────────────────
The chain verifies the ML-DSA-87 signature over the tx body WITH the PQC
extension REMOVED:

- ``B0`` = canonical protobuf bytes of the ``TxBody`` containing the
  messages/memo/timeout but NOT the ``PQCHybridSignature`` extension.
- ``A``  = the ``AuthInfo`` bytes (signer secp256k1 pubkey, SIGN_MODE_DIRECT,
  sequence, fee) — the exact bytes that are broadcast.
- PQC signed message = ``BE32(len(B0)) || B0 || BE32(len(A)) || A`` (4-byte
  big-endian length prefixes; NO hashing, NO domain prefix).
- PQC signature = ``pqc_sign(pqc_secret, message)`` — pure ML-DSA-87, 4627 bytes.
- The ``PQCHybridSignature`` extension is then added to
  ``TxBody.extension_options`` (the CRITICAL extension-options slot) as an
  ``Any`` whose ``type_url`` is ``/qorechain.pqc.v1.PQCHybridSignature`` and
  whose ``value`` is the UTF-8 bytes of the Go-JSON
  ``{"algorithm_id", "pqc_signature", "pqc_public_key"?}`` (standard padded
  base64; ``pqc_public_key`` omitted when not supplied) → final body bytes.
- The CLASSICAL secp256k1 SIGN_MODE_DIRECT signature is computed over
  ``SignDoc(final_body, A, chain_id, account_number)`` and goes in
  ``TxRaw.signatures`` (outside the body). The classical signature never signs
  itself.

The signer's PQC key must be registered on-chain (via ``MsgRegisterPQCKey``)
before hybrid txs PQC-verify — unless ``include_pqc_public_key`` is set, which
embeds the key for auto-registration on first use. Registering the key is the
caller's responsibility.

Determinism note (same caveat as the TS SDK): the BE32 framing is byte-for-byte
deterministic on the wallet side. Cross-implementation determinism (this
``cosmpy`` proto encoding vs. the chain's re-marshal of the same ``TxBody``) is
confirmed for the default bank message types; callers using custom message types
with non-canonical field ordering must ensure their encoding is canonical.
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from typing import Any as TypingAny
from typing import Literal

import httpx
from cosmpy.crypto.keypairs import PrivateKey
from cosmpy.protos.cosmos.bank.v1beta1.tx_pb2 import MsgSend
from cosmpy.protos.cosmos.base.v1beta1.coin_pb2 import Coin
from cosmpy.protos.cosmos.crypto.secp256k1.keys_pb2 import PubKey as Secp256k1PubKey
from cosmpy.protos.cosmos.tx.signing.v1beta1.signing_pb2 import SignMode
from cosmpy.protos.cosmos.tx.v1beta1.tx_pb2 import (
    AuthInfo,
    Fee,
    ModeInfo,
    SignDoc,
    SignerInfo,
    TxBody,
    TxRaw,
)
from google.protobuf.any_pb2 import Any as ProtoAny

from .accounts import Secp256k1Account
from .pqc import (
    ALGORITHM_DILITHIUM5,
    HYBRID_SIG_TYPE_URL,
    PqcKeypair,
    build_hybrid_signature_extension,
    pqc_sign,
)

#: The ``/cosmos.bank.v1beta1.MsgSend`` type URL.
MSG_SEND_TYPE_URL = "/cosmos.bank.v1beta1.MsgSend"

#: Broadcast mode for the REST ``/cosmos/tx/v1beta1/txs`` endpoint.
BroadcastMode = Literal["sync", "async", "block"]

_BROADCAST_MODE_MAP: dict[str, str] = {
    "sync": "BROADCAST_MODE_SYNC",
    "async": "BROADCAST_MODE_ASYNC",
    "block": "BROADCAST_MODE_BLOCK",
}

#: A Cosmos coin as a plain dict, e.g. ``{"denom": "uqor", "amount": "1000"}``.
CoinDict = dict[str, str]
#: A Cosmos ``StdFee``-shaped dict (as produced by :func:`~qorechain.fees.estimate_fee`).
FeeDict = dict[str, TypingAny]


@dataclass(frozen=True)
class BuiltTx:
    """A built, signed transaction plus the intermediate artifacts.

    For a plain :func:`bank_send` the PQC fields are empty; for
    :func:`build_hybrid_tx` they expose the exact bytes signed by ML-DSA-87 so
    the contract can be asserted/audited.
    """

    #: The assembled ``TxRaw`` (final body + authInfo + classical signature).
    tx_raw: TxRaw
    #: Encoded ``TxRaw`` bytes, ready to broadcast.
    tx_raw_bytes: bytes
    #: The ``AuthInfo`` bytes (``A``) — identical in the PQC framing and SignDoc.
    auth_info_bytes: bytes
    #: The exact bytes the ML-DSA-87 signature covered (empty for non-hybrid).
    pqc_signed_message: bytes = b""
    #: The raw ML-DSA-87 signature (Dilithium-5: 4627 bytes; empty if non-hybrid).
    pqc_signature: bytes = b""


def _be32(n: int) -> bytes:
    """A big-endian 4-byte length prefix, matching the chain contract framing."""
    return n.to_bytes(4, "big")


def _to_coins(amount: list[CoinDict]) -> list[Coin]:
    return [Coin(denom=c["denom"], amount=str(c["amount"])) for c in amount]


def _fee_to_proto(fee: FeeDict) -> Fee:
    """Convert an ``estimate_fee``-style dict into a protobuf ``Fee``."""
    amount = _to_coins(list(fee.get("amount", [])))
    gas = int(fee.get("gas", 0))
    granter = fee.get("granter", "") or ""
    payer = fee.get("payer", "") or ""
    return Fee(amount=amount, gas_limit=gas, granter=granter, payer=payer)


def _encode_pubkey_any(compressed_pubkey: bytes) -> ProtoAny:
    """Pack a compressed secp256k1 pubkey into a Cosmos ``Any``."""
    any_pub = ProtoAny()
    any_pub.Pack(Secp256k1PubKey(key=compressed_pubkey), type_url_prefix="/")
    return any_pub


def _build_auth_info_bytes(
    compressed_pubkey: bytes, sequence: int, fee: FeeDict
) -> bytes:
    """Build the single-signer SIGN_MODE_DIRECT ``AuthInfo`` bytes (``A``)."""
    mode_info = ModeInfo(single=ModeInfo.Single(mode=SignMode.SIGN_MODE_DIRECT))
    signer_info = SignerInfo(
        public_key=_encode_pubkey_any(compressed_pubkey),
        mode_info=mode_info,
        sequence=int(sequence),
    )
    auth_info = AuthInfo(signer_infos=[signer_info], fee=_fee_to_proto(fee))
    serialized: bytes = auth_info.SerializeToString()
    return serialized


def _message_parts(message: TypingAny) -> tuple[str, TypingAny]:
    """Extract ``(type_url, value)`` from a :class:`~qorechain.messages.Msg` or dict.

    Accepts either a composer-produced :class:`Msg` (with ``type_url`` /
    ``value`` attributes) or a plain ``{"type_url", "value"}`` dict, so callers
    can pass composer output or hand-built messages interchangeably.
    """
    if isinstance(message, dict):
        return message["type_url"], message["value"]
    return message.type_url, message.value


def _encode_message(message: TypingAny) -> ProtoAny:
    """Encode a message into a protobuf ``Any``.

    ``message`` is a :class:`~qorechain.messages.Msg` or a ``{"type_url",
    "value"}`` dict. ``value`` may be a protobuf message instance (it is
    serialized) or raw ``bytes`` (used verbatim).
    """
    type_url, value = _message_parts(message)
    raw: bytes = (
        value.SerializeToString() if hasattr(value, "SerializeToString") else value
    )
    return ProtoAny(type_url=type_url, value=raw)


def _sign_direct(private_key: bytes, sign_doc: SignDoc) -> bytes:
    """Produce a 64-byte canonical secp256k1 SIGN_MODE_DIRECT signature.

    cosmpy's ``PrivateKey.sign`` signs ``sha256(message)`` deterministically and
    returns the canonical 64-byte ``r || s`` encoding — exactly what Cosmos
    SIGN_MODE_DIRECT expects over the serialized ``SignDoc``.
    """
    priv = PrivateKey(private_key)
    signature: bytes = priv.sign(sign_doc.SerializeToString())
    return signature


def send_messages(
    *,
    account: Secp256k1Account,
    messages: list[TypingAny],
    chain_id: str,
    account_number: int,
    sequence: int,
    fee: FeeDict,
    memo: str = "",
    timeout_height: int = 0,
) -> BuiltTx:
    """Build and sign a tx carrying ANY supported messages (classical only).

    This is the generic counterpart to :func:`bank_send`: it accepts a list of
    composer-produced :class:`~qorechain.messages.Msg` objects (or ``{"type_url",
    "value"}`` dicts), packs each into a Cosmos ``Any``, builds the single-signer
    SIGN_MODE_DIRECT ``AuthInfo`` from the account's compressed secp256k1 pubkey,
    signs the ``SignDoc``, and assembles the ``TxRaw``. It does not broadcast —
    pass :attr:`BuiltTx.tx_raw_bytes` to :func:`broadcast`.

    For a quantum-safe (classical + ML-DSA-87) transaction over the same
    messages, use :func:`build_hybrid_tx` instead.

    :param messages: Messages to include (``Msg`` objects or
        ``{"type_url", "value"}`` dicts). Must be non-empty.
    """
    if not messages:
        raise ValueError("send_messages requires at least one message")

    body = TxBody(
        messages=[_encode_message(m) for m in messages],
        memo=memo,
        timeout_height=int(timeout_height),
    )
    body_bytes = body.SerializeToString()
    auth_info_bytes = _build_auth_info_bytes(account.public_key, sequence, fee)

    sign_doc = SignDoc(
        body_bytes=body_bytes,
        auth_info_bytes=auth_info_bytes,
        chain_id=chain_id,
        account_number=int(account_number),
    )
    signature = _sign_direct(account.private_key, sign_doc)

    tx_raw = TxRaw(
        body_bytes=body_bytes,
        auth_info_bytes=auth_info_bytes,
        signatures=[signature],
    )
    return BuiltTx(
        tx_raw=tx_raw,
        tx_raw_bytes=tx_raw.SerializeToString(),
        auth_info_bytes=auth_info_bytes,
    )


def bank_send(
    *,
    account: Secp256k1Account,
    to_address: str,
    amount: list[CoinDict],
    chain_id: str,
    account_number: int,
    sequence: int,
    fee: FeeDict,
    memo: str = "",
    timeout_height: int = 0,
) -> BuiltTx:
    """Build and sign a bank ``MsgSend`` into a broadcast-ready ``TxRaw``.

    A thin convenience over :func:`send_messages` that constructs
    ``/cosmos.bank.v1beta1.MsgSend`` from ``account.address`` to ``to_address``.
    This does not broadcast — pass :attr:`BuiltTx.tx_raw_bytes` to
    :func:`broadcast`.
    """
    msg = MsgSend(
        from_address=account.address,
        to_address=to_address,
        amount=_to_coins(amount),
    )
    return send_messages(
        account=account,
        messages=[{"type_url": MSG_SEND_TYPE_URL, "value": msg}],
        chain_id=chain_id,
        account_number=account_number,
        sequence=sequence,
        fee=fee,
        memo=memo,
        timeout_height=timeout_height,
    )


def build_hybrid_tx(
    *,
    account: Secp256k1Account,
    pqc_keypair: PqcKeypair,
    messages: list[TypingAny],
    fee: FeeDict,
    chain_id: str,
    account_number: int,
    sequence: int,
    memo: str = "",
    timeout_height: int = 0,
    include_pqc_public_key: bool = False,
) -> BuiltTx:
    """Build a fully signed hybrid (classical + PQC) transaction.

    Follows the chain contract (see the module header). The build sequence:

    1. Encode ``B0`` — the ``TxBody`` WITHOUT the PQC extension.
    2. Encode ``A`` — the single-signer SIGN_MODE_DIRECT ``AuthInfo``.
    3. ``message = BE32(len B0) || B0 || BE32(len A) || A``; ML-DSA-87 sign it.
    4. Build the ``PQCHybridSignature`` extension ``Any`` and attach it to a new
       body identical to step 1 but with ``extension_options = [ext]`` → final
       body bytes.
    5. Classical SIGN_MODE_DIRECT signature over
       ``SignDoc(final_body, A, chain_id, account_number)``.
    6. Assemble ``TxRaw(final_body, A, [classical_sig])``.

    ``messages`` are composer-produced :class:`~qorechain.messages.Msg` objects
    (or ``{"type_url", "value"}`` dicts where ``value`` is a protobuf message or
    raw bytes), so any supported message carries a hybrid signature. The signer's
    PQC key must already be
    registered on-chain via ``MsgRegisterPQCKey`` unless
    ``include_pqc_public_key`` is ``True``.

    :returns: A :class:`BuiltTx` exposing ``pqc_signed_message`` and
        ``pqc_signature`` so the contract can be asserted/audited.
    """
    encoded_messages = [_encode_message(m) for m in messages]

    # 1. B0 — body WITHOUT the PQC extension.
    base_body = TxBody(
        messages=encoded_messages, memo=memo, timeout_height=int(timeout_height)
    )
    b0 = base_body.SerializeToString()

    # 2. A — single-signer AuthInfo (SIGN_MODE_DIRECT).
    auth_info_bytes = _build_auth_info_bytes(account.public_key, sequence, fee)

    # 3. PQC framing + ML-DSA-87 signature over B0 + A (NOT the final body).
    pqc_signed_message = (
        _be32(len(b0)) + b0 + _be32(len(auth_info_bytes)) + auth_info_bytes
    )
    pqc_signature = pqc_sign(pqc_keypair.secret_key, pqc_signed_message)

    # 4. Build the PQC extension Any and attach it to the FINAL body as a
    #    CRITICAL extension option. The Any.value is the Go-JSON of the extension
    #    (standard padded base64; pqc_public_key omitted unless requested).
    ext = build_hybrid_signature_extension(
        ALGORITHM_DILITHIUM5,
        pqc_signature,
        pqc_keypair.public_key if include_pqc_public_key else None,
    )
    ext_any = ProtoAny(
        type_url=HYBRID_SIG_TYPE_URL,
        value=json.dumps(ext, separators=(",", ":")).encode("utf-8"),
    )
    final_body = TxBody(
        messages=encoded_messages,
        memo=memo,
        timeout_height=int(timeout_height),
        extension_options=[ext_any],
    )
    body_bytes_final = final_body.SerializeToString()

    # 5. Classical SIGN_MODE_DIRECT signature over the FINAL body + A.
    sign_doc = SignDoc(
        body_bytes=body_bytes_final,
        auth_info_bytes=auth_info_bytes,
        chain_id=chain_id,
        account_number=int(account_number),
    )
    classical_sig = _sign_direct(account.private_key, sign_doc)

    # 6. Assemble TxRaw.
    tx_raw = TxRaw(
        body_bytes=body_bytes_final,
        auth_info_bytes=auth_info_bytes,
        signatures=[classical_sig],
    )
    return BuiltTx(
        tx_raw=tx_raw,
        tx_raw_bytes=tx_raw.SerializeToString(),
        auth_info_bytes=auth_info_bytes,
        pqc_signed_message=pqc_signed_message,
        pqc_signature=pqc_signature,
    )


def broadcast(
    rest_url: str,
    tx_bytes: bytes,
    *,
    mode: BroadcastMode = "sync",
    timeout: float = 30.0,
    client: httpx.Client | None = None,
) -> TypingAny:
    """Broadcast signed ``TxRaw`` bytes via the REST ``/cosmos/tx/v1beta1/txs``.

    POSTs ``{"tx_bytes": <base64>, "mode": "BROADCAST_MODE_*"}`` and returns the
    decoded JSON response. Broadcasting requires a live node; unit tests mock
    this POST.

    :param mode: ``sync`` (after CheckTx), ``async`` (fire-and-forget), or
        ``block`` (wait for inclusion).
    :raises httpx.HTTPStatusError: On a non-2xx response.
    """
    proto_mode = _BROADCAST_MODE_MAP[mode]
    payload = {
        "tx_bytes": base64.b64encode(tx_bytes).decode("ascii"),
        "mode": proto_mode,
    }
    url = f"{rest_url.rstrip('/')}/cosmos/tx/v1beta1/txs"
    owns_client = client is None
    http = client or httpx.Client(timeout=timeout)
    try:
        resp = http.post(url, json=payload)
        resp.raise_for_status()
        return resp.json()
    finally:
        if owns_client:
            http.close()
